File: UNXNavigatorDemo/Feature/Advanced/Controllers/ViewController03_BackEventIntercept.m
//
//  ViewController03_BackEventIntercept.m
//  UNXNavigatorDemo
//
//  Created by Leo Lee on 2020/10/26.
//

#import <UNXNavigator/UNXNavigator.h>

#import "ViewController03_BackEventIntercept.h"
#import "EventInterceptModel.h"
#import "UIColor+RandomColor.h"

@interface ViewController03_BackEventIntercept () <UNXNavigatorInteractable>

@property (nonatomic, strong) NSArray<EventInterceptModel *> *allModels;
@property (nonatomic, assign) EventInterceptItemType currentItemType;

@end

@implementation ViewController03_BackEventIntercept

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView.tableHeaderView = [[UIView alloc] init];
    self.tableView.tableFooterView = [[UIView alloc] init];
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:NSStringFromClass([UITableViewCell class])];
}

- (UIColor *)unx_shadowImageTintColor {
    return [UIColor customColorWithLightModeColor:^UIColor * _Nonnull{
        return [UIColor lightGrayColor];
    } darkModeColor:^UIColor * _Nonnull{
        return [[UIColor lightGrayColor] colorWithAlphaComponent:0.65];
    }];
}

#pragma mark - Private

- (void)showAlertController {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"提示" message:@"是否继续返回？" preferredStyle:UIAlertControllerStyleAlert];
    [alertController addAction:[UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:NULL]];
    [alertController addAction:[UIAlertAction actionWithTitle:@"确定" style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
        [self.navigationController popViewControllerAnimated:YES];
    }]];
    
    [self presentViewController:alertController animated:YES completion:NULL];
}

#pragma mark - Getter

- (NSArray<EventInterceptModel *> *)allModels {
    if (!_allModels) {
        _allModels = [EventInterceptModel makeAllModels];
    }
    return _allModels;
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.allModels.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:NSStringFromClass([UITableViewCell class])];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:NSStringFromClass([UITableViewCell class])];
    }
    
    EventInterceptModel *item = self.allModels[indexPath.row];
    cell.textLabel.text = item.title;
    cell.accessoryType = item.isSelected ? UITableViewCellAccessoryCheckmark : UITableViewCellAccessoryNone;
    if (@available(iOS 13.0, *)) {
        cell.backgroundColor = [UIColor secondarySystemGroupedBackgroundColor];
    } else {
        cell.backgroundColor = [UIColor customGroupedBackgroundColor];
    }
    
    if (item.isSelected) {
        self.currentItemType = item.itemType;
    }
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    for (EventInterceptModel *item in self.allModels) {
        item.selected = NO;
    }
    
    self.allModels[indexPath.row].selected = YES;
    [tableView reloadData];
}

#pragma mark - UNXNavigatorInteractable

- (BOOL)navigationController:(__kindof UINavigationController *)navigationController willPopViewControllerUsingInteractingGesture:(BOOL)interactingGesture {
    if (self.currentItemType == EventInterceptItemTypeBoth) {
        [self showAlertController];
        return NO;
    }
    
    if (self.currentItemType == EventInterceptItemTypeBackButton) {
        if (!interactingGesture) {
            [self showAlertController];
            return NO;
        }
    }
    
    if (self.currentItemType == EventInterceptItemTypePopGesture) {
        if (interactingGesture) {
            [self showAlertController];
            return NO;
        }
    }
    
    return YES;
}

@end

File: Example/NXNavigationExtensionSwiftDemo/Advanced/ViewController03_BackEventIntercept.swift
//
//  ViewController03_BackEventIntercept.swift
//  NXNavigationExtensionSwiftDemo
//
//  Created by Leo Lee on 2020/11/23.
//

import UIKit
import NXNavigationExtension

class ViewController03_BackEventIntercept: UITableViewController {
    
    private static let reuseIdentifier = String(describing: UITableViewCell.self)
    
    private lazy var items: [EventInterceptItem] = EventInterceptItem.makeAllItems()
    
    private var currentItemType = EventInterceptItem.ItemType.both

    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.tableHeaderView = UIView()
        tableView.tableFooterView = UIView()
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: ViewController03_BackEventIntercept.reuseIdentifier)
    }
    
    override var nx_shadowImageTintColor: UIColor? {
        return UIColor.customDynamic(lightMode: { UIColor.lightGray }, darkMode: { UIColor.lightGray.withAlphaComponent(0.5) })
    }

}

extension ViewController03_BackEventIntercept {
    
    private func showAlertController() {
        let alertController = UIAlertController(title: "提示", message: "是否继续返回？", preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: "取消", style: .cancel, handler: nil))
        alertController.addAction(UIAlertAction(title: "确定", style: .destructive, handler: { [weak self] _ in
            self?.navigationController?.popViewController(animated: true)
        }))
        present(alertController, animated: true, completion: nil)
    }

}

// MARK: - Table view data source

extension ViewController03_BackEventIntercept {

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // #warning Incomplete implementation, return the number of rows
        return items.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let reuseIdentifier = ViewController03_BackEventIntercept.reuseIdentifier
        let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier) ?? UITableViewCell(style: .default, reuseIdentifier: reuseIdentifier)
        
        let item = items[indexPath.row]
        cell.textLabel?.text = item.itemType.rawValue
        cell.accessoryType = item.isSelected ? .checkmark : .none
        
        if #available(iOS 13.0, *) {
            cell.backgroundColor = .secondarySystemGroupedBackground
        } else {
            cell.backgroundColor = UIColor.customGroupedBackground
        }
        
        if item.isSelected { currentItemType = item.itemType }

        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        for item in items { item.isSelected = false }
        items[indexPath.row].isSelected = true
        tableView.reloadData()
    }
    
}

extension ViewController03_BackEventIntercept: NXNavigationExtensionInteractable {
    func navigationController(_ navigationController: UINavigationController, willPopViewControllerUsingInteractingGesture interactingGesture: Bool) -> Bool {
        if currentItemType == .both {
            showAlertController()
            return false
        }
        
        if currentItemType == .backButton {
            if !interactingGesture {
                showAlertController()
                return false
            }
        }
        
        if currentItemType == .popGesture {
            if interactingGesture {
                showAlertController()
                return false
            }
        }
        
        return true
    }
}
